### DISCUSSION TITLE
Removal of objects not in source when only selected folder is published

### DISCUSSION THREAD

**Discussion by TybulOnAzure** [Azure-Player/azure.datafactory.tools#367] [Q&A]
Hi @NowinskiK 

I'm just starting with this great tool and I'm a bit confused about the following case:

1. I have an ADF instance inside which I created a `SomeArea` folder with two pipelines:
![image](https://github.com/Azure-Player/azure.datafactory.tools/assets/139237411/21b8a614-0d0e-4585-a8c2-d36c8c605ce3)
2. I deployed it to live instance using selective approach (remaining options are set to defaults):
```
$selectedObjects = $adf.GetObjectsByFolderName('SomeArea')
$opt.DeleteNotInSource = $true
```
3. Everything works fine so far and those two pipelines have been successfully deployed.
4. Then I removed `pipeline2` from ADF and executed the CI/CD pipeline again.
5. It detected that only `pipeline1` is included in my `SomeArea `folder:
![image](https://github.com/Azure-Player/azure.datafactory.tools/assets/139237411/4de2d87d-6a2f-4dec-ad81-eb2b66dcbc72)
6. ... and deleted one object:
![image](https://github.com/Azure-Player/azure.datafactory.tools/assets/139237411/6e6988aa-da34-41a3-9b64-b1e621dc8b92)

However, both pipelines still exist in the live ADF instance as nothing was actually deleted. So my questions are:
1. What should happen in the described case? Should `pipeline2` be removed from live ADF?
2. What object was marked as removed in logs? 

Basically, I'm trying to think about a scenario, in which multiple teams work on the same ADF instance so they use their own CI/CD pipelines with selective approach. But what if one team would like to remove one of its previously deployed objects - how to achieve this without affecting other teams?

Additional notes: I used `1.7.000` version of ADF Tools.

Piotr Tybulewicz

**Comment by NowinskiK** [comment]
Hi Piotr. Thanks for using the tool and raising a question.
Based on what you have described - yes, `pipeline2` should be deleted. 

However, it's easier way to define deployment using logical folders. 
Use wildcard pattern like this: `*.*@SomeArea`. 
More examples: [Pattern wildcard](https://github.com/Azure-Player/azure.datafactory.tools#pattern-wildcard)

Do you know what object has been deleted in step 6?
If not - enable diagnostic/debug mode (DevOps pipeline) and you should see more details.

The scenario you're considering is vital & real.
`DeleteNotInSource` flag will not drop target objects excluded from the deployment. It's purely excluded from deployment, so should not be touch during the process at all. When you're not sure what will happen with combinations of options - you can check the action matrix presented in section [Selective Deployment](https://github.com/Azure-Player/azure.datafactory.tools#selective-deployment-triggers-and-logic).

Hope this help. If not - feel free to reach me out.

**Reply by TybulOnAzure** [reply]
Hi Kamil.

Thanks for the hint to enable diagnostic mode - everything is clear now:
![image](https://github.com/Azure-Player/azure.datafactory.tools/assets/139237411/06c0e0c0-0073-4c9e-a0c6-019ef5e82fe8)

So the `pipeline2` was not deleted due to `DoNotDeleteExcludedObjects` flag set to `true`. It also means that the summary of the Deleting step is a bit misleading as it doesn't show the number of objects that were actually removed from ADF service, but the number that was considered for removal.

---------------------------------

Anyway, to remove just `pipeline2` I used such approach:
1. I explicitly added it to the list of included objects: 
![image](https://github.com/Azure-Player/azure.datafactory.tools/assets/139237411/e39b0661-f65c-40a1-afd0-a54532fcaa34)
as apparently method to get objects by folder name was not enough as this pipeline doesn't exist in the source code anymore.
2. Then, during the deployment, the pipeline has been removed:
![image](https://github.com/Azure-Player/azure.datafactory.tools/assets/139237411/b9a9ae9d-672b-4525-8daf-db8e9087e403)

Comparison of options:
1. When including objects by folder name, only **pipeline1** was included:
![image](https://github.com/Azure-Player/azure.datafactory.tools/assets/139237411/5113f8d1-3af8-458d-81b1-0de53a29b775)

2. When including objects explicitly by object name, **pipeline2** was also included and therefore was later removed:
![image](https://github.com/Azure-Player/azure.datafactory.tools/assets/139237411/116cd907-901f-43a6-a2ea-d0118c7cb2e0)
